File: python/fidelity-example/fidelity_example.py
import argparse
import asyncio
import sys
from datetime import date
from typing import cast

import zendriver as zd
from fidelity_helper import Fidelity


async def main(argv: list[str]) -> None:
    parser = argparse.ArgumentParser()
    parser.register("type", "date", lambda s: date.fromisoformat(s))  # pyright: ignore[reportUnknownLambdaType, reportUnknownArgumentType]
    _ = parser.add_argument(
        "-a", "--accounts", nargs="+", required=True, help="Account names(s)"
    )
    _ = parser.add_argument(
        "-s",
        "--start",
        type="date",
        required=True,
        help="Start date in YYYY-MM-DD, interpreted in America/New_York",
    )
    _ = parser.add_argument(
        "-e",
        "--end",
        type="date",
        required=True,
        help="End date in YYYY-MM-DD, interpreted in America/New_York",
    )
    args = parser.parse_args(argv)
    driver = await zd.start()
    page = await driver.get("https://www.fidelity.com/")
    fidelity = Fidelity(lambda expr: page.evaluate(expr, await_promise=True))
    _ = input("Login to Fidelity, then press Enter: ")
    ts = await fidelity.get_transactions(
        cast(list[str], args.accounts), cast(date, args.start), cast(date, args.end)
    )
    print(ts)
    await driver.stop()


if __name__ == "__main__":
    x = main
    asyncio.run(main(sys.argv[1:]))

File: node/fidelity-example/src/fidelity-example.ts
import puppeteer from 'puppeteer';
import * as readline from 'readline';
import { Fidelity } from 'fidelity-helper';
import { Command } from 'commander';
import { z } from 'zod';

const program = new Command();

const optionsSchema = z.object({
  accounts: z.array(z.string()),
  start: z.iso.date('Must be valid ISO YYYY-MM-DD date'),
  end: z.iso.date('Must be valid ISO YYYY-MM-DD date'),
}).refine((options) => options.start <= options.end,
  {message: 'start must be less than or equal to end', path: ['start,end']});

program
  .requiredOption('-a, --accounts <str...>', 'Account name(s)')
  .requiredOption('-s, --start <date>', 'Start date in YYYY-MM-DD, interpreted in America/New_York')
  .requiredOption('-e, --end <date>', 'End date in YYYY-MM-DD, interpreted in America/New_York')
  .action(async options => {
    const result = optionsSchema.safeParse(options);
    if (!result.success) {
      result.error.issues.forEach(err => {
        console.error(`${err.path.join('.')}: ${err.message}`);
      });
      process.exit(1);
    }
    const start = new Date(result.data.start + "T00:00:00Z");
    const end = new Date(result.data.end + "T00:00:00Z");
    await getTransactions(result.data.accounts, start, end);
  });

program.parse(process.argv);

async function getTransactions(accounts: string[], start: Date, end: Date) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const browser = await puppeteer.launch({ headless: false });
  const page = await browser.newPage();
  await page.goto('https://www.fidelity.com')
  const fidelity = new Fidelity(async (fn, ...args) => {
    return page.evaluate(fn, ...args);
  });
  // can also simply pass `page.evaluate.bind(page)` instead of wrapping

  await new Promise<void>((resolve) => {
    rl.question('Login to Fidelity, then press Enter: ', () => {
      resolve();
    });
  });

  const transactions = await fidelity.getTransactions(accounts, start, end);
  console.log(`transactions=${JSON.stringify(transactions, null, 2)}`);

  await new Promise<void>((resolve) => {
    rl.question('Press Enter to exit: ', () => {
      rl.close();
      resolve();
    });
  });

  await browser.close();
}
